docs/Execution.md

```diff
@@ -109,3 +109,75 @@ graph TD
         end
     end
 ```
+
+## Custom Execution
+
+MsQuic also supports scenarios where the application layer creates the threads that MsQuic uses to execute on.
+In this mode, the application creates one or more execution contexts that MsQuic will use to run all its internal logic.
+The application is responsible for calling down into MsQuic to allow these execution contexts to run.
+
+To create an execution context, the app much first create an event queue object, which is a platform specific type:
+
+- Windows: IOCP
+- Linux: epoll
+- macOS: kqueue
+
+On Windows, the following types are defined:
+
+```c++
+typedef HANDLE QUIC_EVENTQ;
+
+typedef OVERLAPPED_ENTRY CXPLAT_CQE;
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+(CXPLAT_EVENT_COMPLETION)(
+    _In_ CXPLAT_CQE* Cqe
+    );
+typedef CXPLAT_EVENT_COMPLETION *CXPLAT_EVENT_COMPLETION_HANDLER;
+
+typedef struct CXPLAT_SQE {
+    OVERLAPPED Overlapped;
+    CXPLAT_EVENT_COMPLETION_HANDLER Completion;
+} CXPLAT_SQE;
+```
+
+You will also notice the definiton for `QUIC_SQE` (SQE stands for submission queue entry), which defines the format that all completion events must take so they may be generically processed from the event queue (more on this below).
+
+Once the app has the event queue, it may create the execution context with the `ExecutionCreate` function:
+
+```c++
+HANDLE IOCP = CreateIoCompletionPort(INVALID_HANDLE_VALUE, nullptr, 0, 1);
+QUIC_EXECUTION_CONTEXT_CONFIG ExecConfig = { 0, &IOCP };
+
+QUIC_EXECUTION_CONTEXT* ExecContext = nullptr;
+QUIC_STATUS Status = MsQuic->ExecutionCreate(QUIC_EXECUTION_CONFIG_FLAG_NONE, 0, 1, &ExecConfig, &ExecContext);
+```
+
+The above code createa a new IOCP (for Windows), sets up an execution config, indicating an ideal processor of 0 and the pointer to the IOCP, and then calls MsQuic to create 1 execution context.
```

Comment:
A processor of `0` has no clear, unambiguous meaning on Win32. A [`PROCESSOR_NUMBER`](https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-processor_number) is the only unambiguous processor identifier in Windows user mode, unless MsQuic provides its own canonical representation to apps.

Reply:
One could argue that `0` should always map to `Group 0, Number 0` in any sensible CPU numbering scheme, but certainly once you reach the integer `1`, the point stands.

Reply:
We can look around to see how other platform abstraction layers (std, boost) solve this problem.

Reply:
According to stackoverflow, boost sidesteps the problem by allowing you to get a native thread handle, and then go off and call platform-specific affinity routines yourself. Personally I think an `int` should be a reasonable way to identify CPUs, so if QUIC just exposes ConvertProcessorNumberToProcessorIndex and ConvertProcessorIndexToProcessorNumber helper routines on Windows, you'd be set.

Reply:
Alternatively, you could `typedef int QUIC_CPU_ID` everywhere else and `typedef PROCESSOR_NUMBER QUIC_CPU_ID` on Windows and let apps deal with the difference if they also compile cross-plat.

Reply:
Let's open a bug to track this for possible future changes. For now, it's orthogonal to this PR.

Reply:
IMO it is a blocker for this PR, because it takes an existing internal definition (which we could fix at any point down the road, and was internally consistent within MsQuic and makes it into an API/ABI surface that we cannot break and lacks consistency.

Reply:
This is a "preview" API right now, so we can still change it. We can look at this issue very soon after this PR.